presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RewriteExpressionWithConstantVariable.java

```diff
@@ -0,0 +1,391 @@
+        }
+        return plan;
+    }
+
+    private static class Rewriter
+            extends InternalPlanVisitor<PlanNodeWithConstant, Void>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+
+        public Rewriter(PlanNodeIdAllocator idAllocator)
+        {
+            this.idAllocator = idAllocator;
+        }
+
+        private static boolean isSupportedType(RowExpression rowExpression)
+        {
+            return SUPPORTED_JOIN_KEY_TYPE.contains(rowExpression.getType()) || rowExpression.getType() instanceof VarcharType;
+        }
+
+        private PlanNodeWithConstant accept(PlanNode node)
+        {
+            PlanNodeWithConstant result = node.accept(this, null);
+            return new PlanNodeWithConstant(result.getPlanNode(), result.getConstantExpressionMap());
+        }
+
+        private PlanNodeWithConstant planAndReplace(PlanNode node, boolean passProperties)
+        {
+            List<PlanNodeWithConstant> children = node.getSources().stream().map(this::accept).collect(toImmutableList());
+            PlanNode result = replaceChildren(node, children.stream().map(PlanNodeWithConstant::getPlanNode).collect(toImmutableList()));
+            if (!passProperties) {
+                return new PlanNodeWithConstant(result, ImmutableMap.of());
+            }
+            ImmutableMap.Builder<VariableReferenceExpression, ConstantExpression> properties = ImmutableMap.builder();
+            children.stream().map(PlanNodeWithConstant::getConstantExpressionMap).forEach(properties::putAll);
+            return new PlanNodeWithConstant(result, properties.build());
+        }
+
+        @Override
+        public PlanNodeWithConstant visitPlan(PlanNode node, Void context)
+        {
+            return planAndReplace(node, false);
+        }
+
+        @Override
+        public PlanNodeWithConstant visitFilter(FilterNode node, Void context)
+        {
+            PlanNodeWithConstant rewrittenChild = accept(node.getSource());
+
+            ImmutableMap.Builder<VariableReferenceExpression, ConstantExpression> outputConstantMap = ImmutableMap.builder();
+            RowExpression predicate = node.getPredicate();
+            FilterNode newFilterNode = node;
+            if (!rewrittenChild.getConstantExpressionMap().isEmpty()) {
+                predicate = predicate.accept(new ExpressionRewriter(rewrittenChild.getConstantExpressionMap()), null);
+                newFilterNode = new FilterNode(node.getSourceLocation(), idAllocator.getNextId(), node.getSource(), predicate);
+                outputConstantMap.putAll(rewrittenChild.getConstantExpressionMap());
+            }
+
+            Map<VariableReferenceExpression, ConstantExpression> newConstantMap = new HashMap<>();
+            for (RowExpression conjunct : extractConjuncts(predicate)) {
+                if (conjunct instanceof CallExpression && ((CallExpression) conjunct).getFunctionHandle().getName().equals(EQUAL.getFunctionName().toString())) {
```

Comment:
this looks like a useful utility function we can add to PlannerUtils, something like isEquality